File: android/src/main/java/com/sfpy/reactnative/SFPYCardCaptureView.kt
package com.sfpy.reactnative

import android.widget.FrameLayout
import com.android.safepay.network.error.PaymentError
import com.android.safepay.network.ui.card.SafepayCardCaptureConfiguration
import com.android.safepay.network.ui.card.SafepayCardCaptureController
import com.android.safepay.network.ui.card.SafepayCardCaptureEnvironment
import com.android.safepay.network.ui.card.SafepayCardCaptureResult
import com.android.safepay.network.ui.card.SafepayCardFieldView
import com.facebook.react.uimanager.ThemedReactContext
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class SFPYCardCaptureView(
    private val reactContext: ThemedReactContext,
) : FrameLayout(reactContext) {
    interface Listener {
        fun onReady()
        fun onCardChange(details: SafepayCardFieldView.CardFieldDetails)
        fun onValidated()
        fun onError(code: String?, message: String)
        fun onProceedToAuthentication(result: SafepayCardCaptureResult)
    }

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private val cardFieldView = SafepayCardFieldView(reactContext)

    private var listener: Listener? = null
    private var controller: SafepayCardCaptureController? = null
    private var environment: String? = null
    private var authToken: String? = null
    private var tracker: String? = null
    private var lastConfigurationKey: String? = null
    private var didEmitReady = false

    init {
        addView(
            cardFieldView,
            LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT),
        )

        cardFieldView.setOnCardChangeListener(
            object : SafepayCardFieldView.OnCardChangeListener {
                override fun onCardChange(details: SafepayCardFieldView.CardFieldDetails) {
                    listener?.onCardChange(details)
                }
            },
        )
    }

    fun setListener(listener: Listener?) {
        this.listener = listener
    }

    fun setEnvironment(value: String?) {
        environment = value
        configureControllerIfPossible()
    }

    fun setAuthToken(value: String?) {
        authToken = value
        configureControllerIfPossible()
    }

    fun setTracker(value: String?) {
        tracker = value
        configureControllerIfPossible()
    }

    fun setDisabled(disabled: Boolean) {
        cardFieldView.setDisabled(disabled)
    }

    fun setAutofocus(autofocus: Boolean) {
        if (autofocus) {
            cardFieldView.focus()
        }
    }

    fun submit() {
        val controller = controller
        if (controller == null) {
            listener?.onError(null, "CardCapture is missing required configuration.")
            return
        }

        scope.launch {
            try {
                val result =
                    withContext(Dispatchers.IO) {
                        controller.submit()
                    }
                listener?.onValidated()
                listener?.onProceedToAuthentication(result)
            } catch (error: PaymentError) {
                listener?.onError(error.code, error.msg)
            } catch (error: Exception) {
                listener?.onError(null, error.localizedMessage ?: "Unknown error")
            }
        }
    }

    fun clear() {
        controller?.clear() ?: cardFieldView.clear()
    }

    override fun onDetachedFromWindow() {
        scope.cancel()
        super.onDetachedFromWindow()
    }

    private fun configureControllerIfPossible() {
        val resolvedEnvironment =
            environment
                ?.trim()
                ?.lowercase()
                ?.let {
                    when (it) {
                        "development" -> SafepayCardCaptureEnvironment.DEVELOPMENT
                        "sandbox" -> SafepayCardCaptureEnvironment.SANDBOX
                        "production" -> SafepayCardCaptureEnvironment.PRODUCTION
                        else -> null
                    }
                } ?: return
        val resolvedAuthToken = authToken?.takeIf { it.isNotBlank() } ?: return
        val resolvedTracker = tracker?.takeIf { it.isNotBlank() } ?: return

        val configurationKey = "${resolvedEnvironment.name}|$resolvedAuthToken|$resolvedTracker"
        if (configurationKey == lastConfigurationKey) {
            return
        }

        controller =
            SafepayCardCaptureController(
                configuration =
                    SafepayCardCaptureConfiguration(
                        environment = resolvedEnvironment,
                        authToken = resolvedAuthToken,
                        tracker = resolvedTracker,
                    ),
                cardFieldView = cardFieldView,
            )
        lastConfigurationKey = configurationKey

        if (!didEmitReady) {
            didEmitReady = true
            listener?.onReady()
        }
    }
}

File: ios/SFPYCardCaptureView.swift
import Foundation
import React
import SafepayiOSSDK
import UIKit

@objc(SFPYCardCaptureView)
class SFPYCardCaptureView: UIView {
    @objc var onReady: RCTDirectEventBlock?
    @objc var onCardChange: RCTDirectEventBlock?
    @objc var onValidated: RCTDirectEventBlock?
    @objc var onError: RCTDirectEventBlock?
    @objc var onProceedToAuthentication: RCTDirectEventBlock?

    @objc var environment: NSString? {
        didSet {
            configureControllerIfPossible()
        }
    }

    @objc var authToken: NSString? {
        didSet {
            configureControllerIfPossible()
        }
    }

    @objc var tracker: NSString? {
        didSet {
            configureControllerIfPossible()
        }
    }

    @objc var disabled: Bool = false {
        didSet {
            cardFieldView.isUserInteractionEnabled = !disabled
        }
    }

    @objc var autofocus: Bool = false {
        didSet {
            if autofocus {
                cardFieldView.focus()
            }
        }
    }

    private let cardFieldView = SafepayCardFieldView()
    private var controller: SafepayCardCaptureController?
    private var lastConfigurationKey: String?
    private var didEmitReady = false

    override init(frame: CGRect) {
        super.init(frame: frame)
        setupView()
    }

    required init?(coder: NSCoder) {
        super.init(coder: coder)
        setupView()
    }

    @objc func submit() {
        guard let controller = controller else {
            emitError(code: nil, message: "CardCapture is missing required configuration.")
            return
        }

        Task { [weak self] in
            do {
                let result = try await controller.submit()
                await MainActor.run {
                    self?.onValidated?([:])
                    self?.emitProceed(result)
                }
            } catch let error as PaymentError {
                await MainActor.run {
                    self?.emitError(code: error.code, message: error.message)
                }
            } catch let error as ErrorResponse {
                await MainActor.run {
                    self?.emitError(code: error.code, message: error.errorMessage)
                }
            } catch {
                await MainActor.run {
                    self?.emitError(code: nil, message: error.localizedDescription)
                }
            }
        }
    }

    @objc func clear() {
        controller?.clear() ?? cardFieldView.clear()
    }

    private func setupView() {
        cardFieldView.translatesAutoresizingMaskIntoConstraints = false
        addSubview(cardFieldView)

        NSLayoutConstraint.activate([
            cardFieldView.leadingAnchor.constraint(equalTo: leadingAnchor),
            cardFieldView.trailingAnchor.constraint(equalTo: trailingAnchor),
            cardFieldView.topAnchor.constraint(equalTo: topAnchor),
            cardFieldView.bottomAnchor.constraint(equalTo: bottomAnchor),
        ])

        cardFieldView.cardChangeHandler = { [weak self] details in
            self?.emitCardChange(details)
        }
    }

    private func configureControllerIfPossible() {
        guard
            let environmentValue = environment as String?,
            let environment = parseEnvironment(environmentValue),
            let authTokenValue = authToken as String?,
            !authTokenValue.isEmpty,
            let trackerValue = tracker as String?,
            !trackerValue.isEmpty
        else {
            return
        }

        let configurationKey = "\(environment.rawValue)|\(authTokenValue)|\(trackerValue)"
        guard configurationKey != lastConfigurationKey else {
            return
        }

        controller = SafepayCardCaptureController(
            configuration: SafepayCardCaptureConfiguration(
                environment: environment,
                authToken: authTokenValue,
                tracker: trackerValue
            ),
            cardFieldView: cardFieldView
        )
        lastConfigurationKey = configurationKey

        if !didEmitReady {
            didEmitReady = true
            onReady?([:])
        }
    }

    private func parseEnvironment(_ value: String) -> SafepayCardCaptureEnvironment? {
        SafepayCardCaptureEnvironment(rawValue: value.lowercased())
    }

    private func emitCardChange(_ details: SafepayCardFieldDetails) {
        var card: [String: Any] = [
            "complete": details.complete,
            "validNumber": details.validNumber,
            "validCVC": details.validCVC,
            "validExpiryDate": details.validExpiryDate,
        ]

        card["brand"] = details.brand ?? NSNull()
        card["last4"] = details.last4 ?? ""

        if let expiryMonth = details.expiryMonth {
            card["expiryMonth"] = expiryMonth
        } else {
            card["expiryMonth"] = NSNull()
        }

        if let expiryYear = details.expiryYear {
            card["expiryYear"] = expiryYear
        } else {
            card["expiryYear"] = NSNull()
        }

        onCardChange?(["card": card])
    }

    private func emitError(code: String?, message: String) {
        let errorPayload: [String: Any] = [
            "code": code ?? NSNull(),
            "message": message,
        ]

        onError?(["error": errorPayload])
    }

    private func emitProceed(_ result: SafepayCardCaptureResult) {
        let resultPayload: [String: Any] = [
            "accessToken": result.accessToken,
            "deviceDataCollectionURL": result.deviceDataCollectionURL,
            "cardinalJWT": result.cardinalJWT ?? NSNull(),
            "paymentMethod": [
                "token": result.paymentMethod.token,
                "expirationMonth": result.paymentMethod.expirationMonth,
                "expirationYear": result.paymentMethod.expirationYear,
                "cardTypeCode": result.paymentMethod.cardTypeCode,
                "cardType": result.paymentMethod.cardType,
                "binNumber": result.paymentMethod.binNumber,
                "lastFour": result.paymentMethod.lastFour,
            ],
        ]

        onProceedToAuthentication?(["result": resultPayload])
    }
}
